### PR TITLE
Conviction pallet uses fungible trait

### PR DESCRIPTION
**Pull request by PieWol**
Trying to help out with Issue #226 

- Using MutateFreeze and InspectFreeze as fungible traits for the Currency type in the lib.rs
- should there be renaming from locks to freezes?

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3904006

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3909689

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3910260

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3914646

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3915817

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3916326

File: substrate/frame/conviction-voting/src/lib.rs [hunk 1]
```diff
@@ -107,9 +105,8 @@ pub mod pallet {
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
 		/// Currency type with which voting happens.
-		type Currency: ReservableCurrency<Self::AccountId>
-			+ LockableCurrency<Self::AccountId, Moment = BlockNumberFor<Self>>
-			+ fungible::Inspect<Self::AccountId>;
+		type Currency: MutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>
```
Comment (liamaharon) [inline substrate/frame/conviction-voting/src/lib.rs:108]: and the comment above need to be updated

```suggestion
		type Fungible: MutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>
```

File: substrate/frame/conviction-voting/src/tests.rs [hunk 2]
```diff
@@ -193,7 +194,7 @@ impl Config for Test {
 pub fn new_test_ext() -> sp_io::TestExternalities {
 	let mut t = frame_system::GenesisConfig::<Test>::default().build_storage().unwrap();
 	pallet_balances::GenesisConfig::<Test> {
-		balances: vec![(1, 10), (2, 20), (3, 30), (4, 40), (5, 50), (6, 60)],
+		balances: vec![(1, 11), (2, 21), (3, 31), (4, 41), (5, 51), (6, 61)],
```
Comment (liamaharon) [inline substrate/frame/conviction-voting/src/tests.rs:197]: Why tweak all these?

**Review by liamaharon — COMMENTED** [review]
Looking good, thanks! 

Some notes
- Need to use freezes instead of holds here. Freezes replace locks, while holds replace reserves
- I think you could use the new `Consideration` tickets for reserving funds here, but a benefit of using freezes directly is the minimal code changes that need to take place, so easier review, and the priority is primarily migrating away from Currency so I'm fine with this approach of not using `Consideration` if you prefer that
- Needs a migration to migrate existing locks to freezes

File: substrate/frame/conviction-voting/src/lib.rs (hunk 1 above)
Comment (juangirini) [reply substrate/frame/conviction-voting/src/lib.rs:108]: I would keep `Currency` here, this has already been kept in other migrations like
https://github.com/paritytech/substrate/pull/14020/files#diff-70e9723e9db62816e35f6f885b6770a8449c75a6c2733e9fa7a245fe52c4656cR210-R213

There was a support from @shawntabrizi and @liamaharon towards renaming in here https://github.com/paritytech/substrate/pull/14020#issuecomment-1652250249, but it didn't end happening.

My reasoning behind not renaming is that, at the end of the day, the type is still representing a currency, even though it does not implement a currency trait directly but rather a summary of other traits that put together do represent a currency in this context

File: substrate/frame/conviction-voting/src/lib.rs [hunk 3]
```diff
@@ -291,25 +298,25 @@ pub mod pallet {
...
-		/// - `target`: The account to remove the lock on.
+		/// - `class`: The class of polls to release.
+		/// - `target`: The account to remove the hold on.
 		///
 		/// Weight: `O(R)` with R number of vote of target.
 		#[pallet::call_index(3)]
-		#[pallet::weight(T::WeightInfo::unlock())]
-		pub fn unlock(
+		#[pallet::weight(T::WeightInfo::release())]
+		pub fn release(
```
[the remark this answers is not in the document]
Comment (juangirini) [reply substrate/frame/conviction-voting/src/lib.rs:312]: This shouldn't actually break it https://substrate.stackexchange.com/questions/246/does-renaming-an-extrinsic-require-a-transaction-version-bump

File: substrate/frame/conviction-voting/src/lib.rs (hunk 1 above)
Comment (liamaharon) [reply substrate/frame/conviction-voting/src/lib.rs:108]: Honestly I think it's just confusing and we should rename them all.

> the type is still representing a currency
The type also still represents a fungible. and it's much more descriptive, clear, and less prone to being misinterpreted to call it Fungible.

File: substrate/frame/conviction-voting/src/tests.rs (hunk 2 above)
Comment (PieWol) [reply substrate/frame/conviction-voting/src/tests.rs:197]: absolutely 👍

Comment (PieWol) [reply substrate/frame/conviction-voting/src/tests.rs]: @liamaharon revamped to using freezes now. Still have to update the naming though.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4439879

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4823563

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4836240

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4836204

**Comment by PieWol** [comment]
I've fixed the westend and rococo runtimes so they compile with the pallet now using freezes. Took me quite some time to realize the From impl of the RuntimeFreezeReason for the pallet specific FreezeReason comes from construct_runtime! 😆 

I've used no renaming of imports so I hope this conforms with the latest standardization effort as mentioned in the [tracking-issue](https://github.com/paritytech/polkadot-sdk/issues/226)

Looking forward to some feedback before I start writing a migration :)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4836629

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4836908

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4839504

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4840071

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4872823

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4888825

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5337946

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5338114

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5338270

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5338988

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5339152

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5339597

**Comment by PieWol** [comment]
Hey @ggwpez ,
would you mind giving me a hint why the benchmarks are failing? With my latest changes I was able to make the benchmarks work locally with `cargo t -p pallet-conviction-voting -F runtime-benchmarks.
Your help is much appreciated :)

**Pull request closed without merging**